### PR TITLE
Add stats collection to memory reclaimer

### PR DESCRIPTION
**Pull request by tanjialiang**
During reclaiming there are information that also needs to be collected. Adding the stats collection framework to facilitate reclaiming stats collection.

### CODE REVIEW

File: velox/common/memory/MemoryArbitrator.h [hunk 1]
```diff
@@ -252,6 +256,12 @@ FOLLY_ALWAYS_INLINE std::ostream& operator<<(
 /// through techniques such as disks spilling.
 class MemoryReclaimer {
  public:
+  struct Stats {
+    /// The total number of times of the reclaim attempts that end up failing
+    /// due to reclaiming at non-reclaimable stage.
+    std::atomic<uint64_t> numNonReclaimableAttempts{0};
```
Comment (xiaoxmeng) [inline velox/common/memory/MemoryArbitrator.h:262]: The reclaim is single thread executed so we can just use uint64_t.

File: velox/common/memory/SharedArbitrator.h [hunk 2]
```diff
@@ -200,5 +200,6 @@ class SharedArbitrator : public MemoryArbitrator {
   tsan_atomic<uint64_t> numShrunkBytes_{0};
   tsan_atomic<uint64_t> numReclaimedBytes_{0};
   tsan_atomic<uint64_t> reclaimTimeUs_{0};
+  MemoryReclaimer::Stats reclaimerStats_;
```
Comment (xiaoxmeng) [inline velox/common/memory/SharedArbitrator.h:203]:
```
tsan_atomic<uint64_t> nonReclaimableAttempts_{0};
```

File: velox/common/memory/SharedArbitrator.cpp [hunk 3]
```diff
@@ -404,7 +404,7 @@ uint64_t SharedArbitrator::reclaim(
     try {
       freedBytes = pool->shrink(targetBytes);
       if (freedBytes < targetBytes) {
-        pool->reclaim(targetBytes - freedBytes);
+        pool->reclaim(targetBytes - freedBytes, reclaimerStats_);
```
Comment (xiaoxmeng) [inline velox/common/memory/SharedArbitrator.cpp:407]: We can pass the stats as a local variable here?

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang LGTM. Thanks!

File: velox/common/memory/tests/MemoryArbitratorTest.cpp [hunk 4]
```diff
@@ -312,9 +314,9 @@ TEST_F(MemoryReclaimerTest, default) {
       uint64_t reclaimableBytes;
       ASSERT_FALSE(pool->reclaimableBytes(reclaimableBytes));
       ASSERT_EQ(reclaimableBytes, 0);
-      ASSERT_EQ(pool->reclaim(0), 0);
-      ASSERT_EQ(pool->reclaim(100), 0);
-      ASSERT_EQ(pool->reclaim(kMaxMemory), 0);
+      ASSERT_EQ(pool->reclaim(0, stats_), 0);
+      ASSERT_EQ(pool->reclaim(100, stats_), 0);
+      ASSERT_EQ(pool->reclaim(kMaxMemory, stats_), 0);
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/MemoryArbitratorTest.cpp:319]: Can you check the stats_ are all zero?

File: velox/common/memory/tests/MemoryArbitratorTest.cpp [hunk 5]
```diff
@@ -591,15 +595,16 @@ TEST_F(MemoryReclaimerTest, orderedReclaim) {
   // child and 4 from 4th child and 1 from 2nd.
   // So expected reclaimable allocation units are {0, 0, 2, *0*, *0*}
   ASSERT_EQ(
-      root->reclaimer()->reclaim(root.get(), 10 * allocUnitBytes),
+      root->reclaimer()->reclaim(root.get(), 10 * allocUnitBytes, stats_),
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/MemoryArbitratorTest.cpp:598]: Can we have a fake test here to verify the stats has been bumped and aggregated? Thanks!

File: velox/exec/OrderBy.h [hunk 6]
```diff
@@ -57,7 +57,8 @@ class OrderBy : public Operator {
     return finished_;
   }
 
-  void reclaim(uint64_t targetBytes) override;
+  void reclaim(uint64_t targetBytes, memory::MemoryReclaimer::Stats& stats)
```
Comment (xiaoxmeng) [inline velox/exec/OrderBy.h:60]: Can you update stats in those operator reclaimer? Thanks!

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 7]
```diff
@@ -2147,6 +2150,60 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, driverInitTriggeredArbitration) {
...
+    });
+    allocation0.buffer = op->pool()->allocate(16 * MB);
+    allocation0.pool = op->pool();
+    allocation0.size = 16 * MB;
+
+    allocation1.buffer = op->pool()->allocate(16 * MB);
+    allocation1.pool = op->pool();
+    allocation1.size = 16 * MB;
+
+    TestAllocation ret;
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2184]:
```
return TestAllocation{}
```

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 8]
```diff
@@ -2147,6 +2150,60 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, driverInitTriggeredArbitration) {
...
+                                               MemoryReclaimer::Stats& stats) {
+    ++stats.numNonReclaimableAttempts;
+    outerAlloc.free();
+    return true;
+  });
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  AssertQueryBuilder(duckDbQueryRunner_)
+      .queryCtx(queryCtx)
+      .plan(PlanBuilder(planNodeIdGenerator, pool())
+                .values(vectors)
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2198]: Can you test with multiple drivers to see if they are aggregated? Thanks!

File: velox/common/memory/tests/SharedArbitratorTest.cpp [hunk 9]
```diff
@@ -2147,6 +2150,60 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, driverInitTriggeredArbitration) {
...
+  std::shared_ptr<core::QueryCtx> queryCtx = newQueryCtx(32 * MB);
+  ASSERT_EQ(queryCtx->pool()->capacity(), 0);
+  ASSERT_EQ(queryCtx->pool()->maxCapacity(), 32 * MB);
+  auto additionalCtxLeafPool = queryCtx->pool()->addLeafChild("ctx_leaf");
+  TestAllocation outerAlloc;
+  outerAlloc.buffer = additionalCtxLeafPool->allocate(8 * MB);
+  outerAlloc.pool = additionalCtxLeafPool.get();
+  outerAlloc.size = 8 * MB;
+  fakeOperatorFactory_->setAllocationCallback([&](Operator* op) {
+    TestAllocation allocation0;
```
Comment (xiaoxmeng) [inline velox/common/memory/tests/SharedArbitratorTest.cpp:2170]: you might want to use an atomic bool to make sure we trigger arbitration reclaim once with multiple drivers?

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang thanks for the update!

File: velox/common/memory/MemoryArbitrator.cpp [hunk 10]
```diff
@@ -243,6 +244,17 @@ void MemoryReclaimer::abort(MemoryPool* pool, const std::exception_ptr& error) {
   });
 }
 
+bool MemoryReclaimer::Stats::operator==(
+    const MemoryReclaimer::Stats& other) const {
+  return std::tie(numNonReclaimableAttempts) ==
```
Comment (xiaoxmeng) [inline velox/common/memory/MemoryArbitrator.cpp:249]: Don't need for std::tie as there is only one element

File: velox/exec/tests/AggregationTest.cpp [hunk 11]
```diff
@@ -2045,14 +2046,17 @@ DEBUG_ONLY_TEST_F(AggregationTest, reclaimDuringInputProcessing) {
...
+          reclaimerStats_);
       // The hash table itself in the grouping set is not cleared so it still
       // uses some memory.
       ASSERT_LT(op->pool()->currentBytes(), usedMemory);
     } else {
       VELOX_ASSERT_THROW(
           op->reclaim(
-              folly::Random::oneIn(2) ? 0 : folly::Random::rand32(rng_)),
+              folly::Random::oneIn(2) ? 0 : folly::Random::rand32(rng_),
+              reclaimerStats_),
```
Comment (xiaoxmeng) [inline velox/exec/tests/AggregationTest.cpp:2059]: Can you add check stats for these tests to improve coverage? thanks!

File: velox/exec/tests/OrderByTest.cpp [hunk 12]
```diff
@@ -216,6 +216,7 @@ class OrderByTest : public OperatorTestBase {
   }
 
   folly::Random::DefaultGenerator rng_;
+  memory::MemoryReclaimer::Stats reclaimerStats_;
```
Comment (xiaoxmeng) [inline velox/exec/tests/OrderByTest.cpp:219]: Dittos

File: velox/exec/tests/HashJoinTest.cpp [hunk 13]
```diff
@@ -845,6 +845,7 @@ class HashJoinTest : public HiveConnectorTestBase {
   RowTypePtr probeType_;
   RowTypePtr buildType_;
 
+  memory::MemoryReclaimer::Stats reclaimerStats_;
```
Comment (xiaoxmeng) [inline velox/exec/tests/HashJoinTest.cpp:848]: ditto

File: velox/common/memory/tests/SharedArbitratorTest.cpp (hunk 7 above)
Comment (xiaoxmeng) [reply velox/common/memory/tests/SharedArbitratorTest.cpp:2185]: You didn't change?

**Review by xiaoxmeng — COMMENTED** [review]
@tanjialiang thanks for the update!

File: velox/common/memory/MemoryArbitrator.cpp (hunk 10 above)
Comment (tanjialiang) [reply velox/common/memory/MemoryArbitrator.cpp:249]: For later extension. But yeah we can just compare the single variable for now.

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang LGTM. We could consider to add actual reclaim time and reclaim wait time in this stats to tell how much time spent on spilling and how much time spent on waiting for task to pause.

File: velox/exec/tests/HashJoinTest.cpp [hunk 14]
```diff
@@ -5538,6 +5557,7 @@ DEBUG_ONLY_TEST_F(HashJoinTest, reclaimDuringWaitForProbe) {
   task.reset();
 
   taskThread.join();
+  ASSERT_EQ(reclaimerStats_, memory::MemoryReclaimer::Stats{});
```
Comment (xiaoxmeng) [inline velox/exec/tests/HashJoinTest.cpp:5560]: Consider to provide empty() in followup. Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@tanjialiang thanks for the update!

**Pull request closed without merging**